core/src/main/java/org/mskcc/cbio/portal/util/GenePanelUtil.java

```diff
@@ -0,0 +1,165 @@
+        DaoGeneOptimized daoGeneOptimized = DaoGeneOptimized.getInstance();
+        for (String panelGene : genes) {
+            try {
+                long geneId = Long.parseLong(panelGene);
+                CanonicalGene canonicalGene = daoGeneOptimized.getGene(geneId);
+                if (canonicalGene != null) {
+                    canonicalGenes.add(canonicalGene);
+                }
+                else {
+                    ProgressMonitor.logWarning("Could not find gene in the database: " + geneId);
+                }
+            }
+            catch (NumberFormatException e) {
+                List<CanonicalGene> canonicalGenesList = daoGeneOptimized.getGene(panelGene, true);
+                if (canonicalGenesList != null && !canonicalGenesList.isEmpty()) {
+                    // we do not want multiple genes added to the gene panel object
+                    // for a single gene symbol found in the data file 
+                    canonicalGenes.add(canonicalGenesList.get(0));
+                }
+                else {
+                    ProgressMonitor.logWarning("Could not find gene in the database: " + panelGene);
+                }
+            }
+        }
+
+        return (canonicalGenes.size() == genes.length) ? canonicalGenes : null;
+    }
+
+    /**
+     * Gets a Pair representing the sets of genes to add and to remove from a gene
+     * panel in the database.
+     */
+    public static Pair getAddRemove(Set<CanonicalGene> incoming, Set<CanonicalGene> original) {
+        // A -> incoming, B -> original
+        Set<CanonicalGene> toAdd = new HashSet<>();
+        Set<CanonicalGene> toRemove = new HashSet<>();
+        Set<CanonicalGene> AnB = intersection(new ArrayList<>(incoming), new ArrayList<>(original));
+        Set<CanonicalGene> AuB = union(new ArrayList<>(incoming), new ArrayList<>(original));
+
+        // 1. Is A == B?
+        if (AnB.equals(AuB)) {
+            return new Pair(toAdd, toRemove);
+        }
+
+        // 2. Is A ^ B == nil?
+        else if (AnB.isEmpty()) {
+            // Remove B
+            toRemove = original;
+            // Add A
+            toAdd = incoming;
+        }
+
+        // 3. Is AnB == B?
+        else if (AnB.equals(original)) {
+            // Add A - B
+            toAdd = difference(new ArrayList<>(incoming), new ArrayList<>(original));
+        }
+
+        // 4. Is AnB == A?
+        else if (AnB.equals(incoming)) {
+            // Remove B - A
+            toRemove = difference(new ArrayList<>(original), new ArrayList<>(incoming));
+        }
+
+        // 5. Intersect but neither are supersets of one other
+        else {
+            // Add A - AnB
+            toAdd = difference(new ArrayList<>(incoming), new ArrayList<>(AnB));
+            // Remove B - AnB
+            toRemove = difference(new ArrayList<>(original), new ArrayList<>(AnB));
+        }
+
+        return new Pair(toAdd, toRemove);
+    }
```

Comment:
This seems too complex for what it was trying to do. I replaced it with this and it passed all the tests:
```suggestion
    public static Pair getAddRemove(Set<CanonicalGene> incoming, Set<CanonicalGene> original) {
        Set<CanonicalGene> add = incoming.stream()
            .filter(e -> !original.contains(e))
            .collect(Collectors.toSet());
        Set<CanonicalGene> remove = original.stream()
            .filter(e -> !incoming.contains(e))
            .collect(Collectors.toSet());
        
        return new Pair(add, remove);
    }
```

Are these functionally equivalent, or am I missing something?

Reply:
I never thought of it that way... I think they are functionally equivalent!